source/AI.cpp

```diff
@@ -309,6 +312,199 @@ AI::AI(const List<Ship> &ships, const List<Minable> &minables, const List<Flotsa
 
 
 // Fleet commands from the player.
+void AI::IssueFormationChange(const PlayerInfo &player)
+{
+	// Figure out what ships we are giving orders to
+	vector<Ship *> ships;
+	bool fullFleet = player.SelectedShips().empty();
+	if(fullFleet)
+	{
+		for(const shared_ptr<Ship> &it : player.Ships())
+			if(it.get() != player.Flagship() && !it->IsParked()
+				&& !it->IsDestroyed())
+				ships.push_back(it.get());
+	}
+	else
+		for(const weak_ptr<Ship> &it : player.SelectedShips())
+		{
+			shared_ptr<Ship> ship = it.lock();
+			if(ship)
+				ships.push_back(ship.get());
+		}
+	
+	// This should never happen, but just in case:
+	if(ships.empty())
+	{
+		if(fullFleet)
+			Messages::Add("No ships in the fleet to change formation for.", Messages::Importance::High);
+		else
+			Messages::Add("No ships selected that can change formation.", Messages::Importance::High);
+		
+		return;
+	}
+	
+	const Set<FormationPattern> &formationPatterns = GameData::Formations();
+	
+	// First check which and how many formations we have in the current set of selected ships.
+	const FormationPattern *toSet = nullptr;
+	bool moreFound = false;
+	for(Ship *ship : ships)
+	{
+		const FormationPattern *shipsPattern = ship->GetFormationPattern();
+		if(shipsPattern)
+		{
+			if(!toSet)
+				toSet = formationPatterns.Find(shipsPattern->Name());
+			else if(toSet->Name() != shipsPattern->Name())
+				moreFound = true;
+		}
+	}
+	
+	// Now determine what formation to set.
+	// If more than one formation was found in the set, then we set it to the first one found.
+	if(!toSet)
+	{
+		// If no formation was set at all, then we set the first one from the set of formations.
+		auto it = formationPatterns.begin();
+		if(it != formationPatterns.end())
+			toSet = formationPatterns.Find(it->first);
+	}
+	else if(!moreFound)
```

Comment:
I feel `else if` is correct here, there are actually 3 options:
- toSet is nullptr (no existing pattern was found at all on all the ships)
- moreFound is false (only one pattern was found on all ships)
- moreFound is true (more than one pattern was found on all ships)

If moreFound is true, then what should happen is that all ships get the same first-found pattern. This option doesn't need any code here, since all variables are already set to just apply the pattern change. I will add some comments to show this 3rd option more explicitly.